consul/consul.go

```diff
@@ -63,58 +56,71 @@ func (c *Consul) getServicesUsingProviderWithRetriesOnAgentFailure(provide Servi
 	return nil, fmt.Errorf("An error occurred getting services from Consul. Giving up")
 }
 
-func (c *Consul) getServicesUsingAgent(name string, agent *consulapi.Client) ([]*consulapi.CatalogService, error) {
+func (c *Consul) getServicesUsingAgent(name string, agent *consulapi.Client) ([]*service.Service, error) {
 	datacenters, err := agent.Catalog().Datacenters()
 	if err != nil {
 		return nil, err
 	}
-	var allServices []*consulapi.CatalogService
+	var allServices []*service.Service
 
 	for _, dc := range datacenters {
 		dcAwareQuery := &consulapi.QueryOptions{
 			Datacenter: dc,
 		}
-		services, _, err := agent.Catalog().Service(name, c.config.Tag, dcAwareQuery)
+		allConsulServices, _, err := agent.Catalog().Service(name, c.config.Tag, dcAwareQuery)
 		if err != nil {
 			return nil, err
 		}
-		allServices = append(allServices, services...)
+		for _, consulService := range allConsulServices {
+			allServices = append(allServices, c.consulServiceToService(consulService))
+		}
 	}
 	return allServices, nil
 }
 
-func (c *Consul) GetAllServices() ([]*consulapi.CatalogService, error) {
+func (c *Consul) GetAllServices() ([]*service.Service, error) {
 	return c.getServicesUsingProviderWithRetriesOnAgentFailure(c.getAllServices)
 }
 
-func (c *Consul) getAllServices(agent *consulapi.Client) ([]*consulapi.CatalogService, error) {
+func (c *Consul) getAllServices(agent *consulapi.Client) ([]*service.Service, error) {
 	datacenters, err := agent.Catalog().Datacenters()
 	if err != nil {
 		return nil, err
 	}
-	var allInstances []*consulapi.CatalogService
+	var allInstances []*service.Service
 
 	for _, dc := range datacenters {
 		dcAwareQuery := &consulapi.QueryOptions{
 			Datacenter: dc,
 		}
-		services, _, err := agent.Catalog().Services(dcAwareQuery)
+		consulServices, _, err := agent.Catalog().Services(dcAwareQuery)
 		if err != nil {
 			return nil, err
 		}
-		for service, tags := range services {
+		for consulService, tags := range consulServices {
 			if contains(tags, c.config.Tag) {
-				serviceInstances, _, err := agent.Catalog().Service(service, c.config.Tag, dcAwareQuery)
+				consulServiceInstances, _, err := agent.Catalog().Service(consulService, c.config.Tag, dcAwareQuery)
 				if err != nil {
 					return nil, err
 				}
-				allInstances = append(allInstances, serviceInstances...)
+				for _, consulServiceInstance := range consulServiceInstances {
+					allInstances = append(allInstances, c.consulServiceToService(consulServiceInstance))
+				}
 			}
 		}
 	}
 	return allInstances, nil
 }
 
+func (c *Consul) consulServiceToService(consulService *consulapi.CatalogService) *service.Service {
```

Comment:
How about moving this function to `service` package and make it constructor function e.g. `func FromConsul(consulService *consulapi.CatalogService) *service.Service`